### PR TITLE
src/crimson: Add support for the OSD to receive config changes

### PR DESCRIPTION
**Pull request by amathuria**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/mgr/client.cc [hunk 1]
```diff
@@ -168,6 +170,26 @@ void Client::report()
...
+    logger().warn("cannot send report; no conn available");
+    return seastar::now();
+  }
+  auto report = make_message<MMgrReport>();
+  if (daemon_name.size()) {
+    report->daemon_name = daemon_name;
+  } else {
+    report->daemon_name = local_conf()->name.get_id();
+  }
+  report->service_name = service_name;
```
Comment (Matan-B) [inline src/crimson/mgr/client.cc:186]: service_name will always remain empty for now, although it seems that handle_report expects that.

File: src/crimson/mgr/client.cc [hunk 2]
```diff
@@ -168,6 +170,26 @@ void Client::report()
   });
 }
 
+void Client::_send_report()
+{
+  gate.dispatch_in_background(__func__, *this, [this] {
+  if (!conn) {
```
Comment (Matan-B) [inline src/crimson/mgr/client.cc:176]: nit: indentation

File: src/mgr/DaemonServer.cc [hunk 3]
```diff
@@ -699,8 +699,10 @@ bool DaemonServer::handle_report(const ref_t<MMgrReport>& m)
     ceph_assert(daemon != nullptr);
     {
       std::lock_guard l(daemon->lock);
-      auto &daemon_counters = daemon->perf_counters;
-      daemon_counters.update(*m.get());
+      if (m->declare_types.size() > 0 || m->undeclare_types.size() > 0) {
```
[the remark this answers is not in the document]
Comment (amathuria) [reply src/mgr/DaemonServer.cc:702]: Yes I can make that change

**Review by athanatos — APPROVED** [review]
LGTM other than @Matan-B 's comments

File: src/crimson/mgr/client.cc [hunk 4]
```diff
@@ -168,6 +170,34 @@ void Client::report()
...
+    if (!conn) {
+      logger().warn("cannot send report; no conn available");
+      return seastar::now();
+    }
+    auto report = make_message<MMgrReport>();
+    // Adding empty information since we don't support perfcounters yet
+    report->undeclare_types.emplace_back();
+    ENCODE_START(1, 1, report->packed);
+    report->declare_types.emplace_back();
+    ENCODE_FINISH(report->packed);
```
Comment (amathuria) [inline src/crimson/mgr/client.cc:187]: @Matan-B addressed your review comment to avoid changing classical code. Does this look okay to you?

Comment (Matan-B) [reply src/crimson/mgr/client.cc:187]: Yes, we will have to verify that the mgr is ok with empty stats, haven't seen any asserts so it should be fine.

File: src/crimson/mgr/client.cc [hunk 5]
```diff
@@ -168,6 +170,34 @@ void Client::report()
...
+
+    if (daemon_name.size()) {
+    report->daemon_name = daemon_name;
```
Comment (Matan-B) [inline src/crimson/mgr/client.cc:190]: indentation

File: src/crimson/mgr/client.cc [hunk 6]
```diff
@@ -168,6 +170,34 @@ void Client::report()
...
+    } else {
+    report->daemon_name = local_conf()->name.get_id();
```
Comment (Matan-B) [inline src/crimson/mgr/client.cc:192]: ditto

**Review by Matan-B — APPROVED** [review]
LGTM

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-01-18_13:55:25-crimson-rados-wip-matanb-crimson-testing-18.1-crimson-only-distro-crimson-smithi/

**Pull request merged**